impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java

```diff
@@ -347,9 +284,14 @@ public synchronized void removeBroker(int nodeId) throws IllegalArgumentExceptio
     public synchronized void close() throws Exception {
         try {
             try {
-                // shutdown controller last seems to be significant to the kraft case.
+                // with kraft, if we don't shut down the controller last, we sometimes see a hang.
+                // https://issues.apache.org/jira/browse/KAFKA-14287
                 servers.entrySet().stream()
-                        .sorted(Map.Entry.comparingByKey(Comparator.reverseOrder()))
+                        .filter(e -> !portsAllocator.containsPort(Listener.CONTROLLER, e.getKey()))
+                        .map(Map.Entry::getValue)
+                        .forEach(Server::shutdown);
+                servers.entrySet().stream()
+                        .filter(e -> portsAllocator.containsPort(Listener.CONTROLLER, e.getKey()))
```

Comment:
I think this filter condition is extraneous as calling shutdown on the non controllers is safe (AFICT).

Reply:
I wasn't so sure about the idempotency in the kraft case.   If I invoke `kafka.server.KafkaRaftServer#shutdown` twice, what stops the second invocation overtaking the first?   In our case it would be safe as we are calling our shutdowns sequentially from a single thread.  To be safe I'd prefer to keep it this way.